junit-jupiter-api/src/main/java/org/junit/jupiter/api/AssertThrows.java

```diff
@@ -42,4 +45,25 @@
 			String.format("Expected %s to be thrown, but nothing was thrown.", getCanonicalName(expectedType)));
 	}
 
+	@SuppressWarnings("unchecked")
+	static <T extends Throwable> T assertThrows(Class<T> expectedType, Executable executable,
+			Supplier<String> messageSupplier) {
+		return assertThrows(expectedType, executable, nullSafeGet(messageSupplier));
+	}
+
+	@SuppressWarnings("unchecked")
+	static <T extends Throwable> T assertThrows(Class<T> expectedType, Executable executable, String message) {
+		try {
+			executable.execute();
+		}
+		catch (Throwable actualException) {
+			if (expectedType.isInstance(actualException)) {
+				return (T) actualException;
+			}
+			else {
+				throw new AssertionFailedError(message, actualException);
```

Comment:
The user-supplied message should be included in the failure message in addition to the standard failure message, not as a replacement for the standard failure message.

Please consult other assertion implementations for inspiration.